js/ui/appDisplay.js

```diff
@@ -2743,3 +2780,80 @@ class AppCenterIcon extends AppIcon {
         return true;
     }
 });
+
+const HackAppIconState = {
+    ACTIVATED: ViewIconState.NUM_STATES + 2,
+    DEACTIVATED: ViewIconState.NUM_STATES + 3
+};
+
+var HackAppIcon = GObject.registerClass(
+class HackAppIcon extends ViewIcon {
+    _init(parentView) {
+        let viewIconParams = { isDraggable: false,
+                               showMenu: false,
+                               parentView: parentView };
+
+        let buttonParams = { button_mask: St.ButtonMask.ONE | St.ButtonMask.TWO };
+        let iconParams = { createIcon: this._createIcon.bind(this) };
+
+
+        super._init(viewIconParams, buttonParams, iconParams);
+
+        let activated = global.settings.get_boolean('hack-mode-enabled');
+        this.iconState = activated ? HackAppIconState.ACTIVATED : HackAppIconState.DEACTIVATED;
+
+        this._hackModeId = global.settings.connect('changed::hack-mode-enabled', () => {
+            let activated = global.settings.get_boolean('hack-mode-enabled');
+            this.iconState = activated ? HackAppIconState.ACTIVATED : HackAppIconState.DEACTIVATED;
+        });
+
+        this.canDrop = false;
+
+        this._iconContainer = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                              x_expand: true, y_expand: true });
+        this._iconContainer.add_child(this.icon.actor);
+
+        this.actor.set_child(this._iconContainer);
+    }
+
+    _createIcon(iconSize) {
+        let iconUri = 'resource:///org/gnome/shell/theme/hack-button-on.png';
+        if (this.iconState === HackAppIconState.DEACTIVATED)
+            iconUri = 'resource:///org/gnome/shell/theme/hack-button-off.png';
+
+        let iconFile = Gio.File.new_for_uri(iconUri);
+        let gicon = new Gio.FileIcon({ file: iconFile });
+
+        return new St.Icon({ gicon: gicon,
```

Comment:
minor: can be shortened, `return new St.Icon({gicon, iconSize});`

Reply:
the `iconSize` is not the key for the dict for `St.Icon`, the key is `icon_size`.

Reply:
Oh, I misread that. But by coincidence, my comment could still apply :smile: Both `icon_size` and `iconSize` are accepted for GObject property names, but I see that if you're using underscores consistently then using the shorthand here would be confusing.